### ISSUE TITLE
`optipng`: Segmentation Fault in master 

### ISSUE THREAD

**Issue by the-kenny** [NixOS/nixpkgs#28106]
## Issue description

Optipng crashes with a segmentation fault on current `master` as of c2ef89aac97b6eb8e3efe53173ec8eea7d2ebf17.

`gdb` points to the `cexcept` library used to throw/catch exceptions. https://github.com/gaspardle/OptiPNG.net/blob/master/optipng/src/cexcept/cexcept.h

```
Starting program: /nix/store/alplyfp3sz62qs87rlmwpl6mbk25p27r-optipng-0.7.6/bin/optipng

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff73e2d70 in __longjmp_chk () from /nix/store/2jyv8gb7paxxmc31h47ky8my18bd0b85-glibc-2.25/lib/libc.so.6
(gdb) bt
#0  0x00007ffff73e2d70 in __longjmp_chk () from /nix/store/2jyv8gb7paxxmc31h47ky8my18bd0b85-glibc-2.25/lib/libc.so.6
#1  0x0000000000402560 in opng_check_idat_size.part ()
```
relevant source: https://github.com/gaspardle/OptiPNG.net/blob/5c36a086e128b6e50317aeef5395ae7e95446e12/optipng/src/optipng/optim.c#L403-L408

### Steps to reproduce

`nix-build . -A optipng && result/bin/optipng`

## Bisect

Automatic `git-bisect` points to 093cc00cdd9d8cf31ecce5bc1dd3645c460a1b98 being the first bad revision. However, I don't have enough knowledge of this piece of *nix to understand why it causes optipng to fail in the `longjmp`.

## Technical details

* System: nixos-master
* Nix version: 1.11.11
* Nixpkgs version: c2ef89aac97b6eb8e3efe53173ec8eea7d2ebf17
* Sandboxing enabled: yes

CC @Ericson2314

**Comment by the-kenny** [comment]
Reading into the causing commit's comments made me try 

```
  preConfigure = ''
    export LD=$CC
  '';
```

which seems to work around the issue. However, I'm hesitant to commit this as a fix without understanding the issue here.

**Comment by Ericson2314** [comment]
@the-kenny So some packages use the C compiler instead of linker directly to link. Nixpkgs's cc-wrapper is aware of this, for example, with separate `NIX_CFLAGS_LINK` and `NIX_LDFLAGS` environment variables.

Classic make builtin rules however only have `CC`, `CFLAGS`, `LD`, and `LDFLAGS`, so then either in a configure script or make itself they seem to *default* `LD` to `CC`, but manually specifying `LD` breaks this. Manually specifying `LD=$CC`, however, avoids that pitfall.

**Comment by cstrahan** [comment]
Here's a blurb from optipng's `configure` script:

```bash
sed_config="
    s|@prefix@|$prefix|g
    s|@exec_prefix@|$exec_prefix|g
    s|@bindir@|$bindir|g
    s|@mandir@|$mandir|g
    s|@man1dir@|$man1dir|g
    s|@USE_SYSTEM_LIBPNG_TRUE@|$USE_SYSTEM_LIBPNG_TRUE|g
    s|@USE_SYSTEM_LIBPNG_FALSE@|$USE_SYSTEM_LIBPNG_FALSE|g
    s|@USE_SYSTEM_ZLIB_TRUE@|$USE_SYSTEM_ZLIB_TRUE|g
    s|@USE_SYSTEM_ZLIB_FALSE@|$USE_SYSTEM_ZLIB_FALSE|g
    s|@CC@|${CC-cc}|g
    s|@CFLAGS@|${CFLAGS--O}|g
    s|@CPP@|${CPP-\$(CC) -E}|g
    s|@CPPFLAGS@|${CPPFLAGS-}|g
    s|@LD@|${LD-\$(CC)}|g
    s|@LDFLAGS@|${LDFLAGS--s}|g
    s|@AR@|${AR-ar}|g
    s|@ARFLAGS@|${ARFLAGS-cru}|g
    s|@RANLIB@|${RANLIB-ranlib}|g
    s|@LIBS@|${LIBS-}|g
    s|@LIBM@|${LIBM--lm}|g
    s|@LIBZ@|${LIBZ--lz}|g
    s|@LIBPNG@|${LIBPNG--lpng}|g
    s|@DIFF@|${DIFF-diff -b -u}|g
    s|@RM_F@|${RM_F-rm -f}|g
    s|@LIBPNG_MK@|${LIBPNG_MK-Makefile}|g
    s|@LIBPNG_MK_DEF@|${LIBPNG_MK_DEF-PNGLIBCONF_H_PREBUILT=pnglibconf.h.optipng}|g
    s|@LIBPNG_DISTCLEAN@|${LIBPNG_DISTCLEAN-distclean}|g
    s|@LIBPNG_DISTCLEAN_XCMD@|${LIBPNG_DISTCLEAN_XCMD-true}|g
    s|@ZLIB_MK@|${ZLIB_MK-Makefile}|g
    s|@ZLIB_DISTCLEAN@|${ZLIB_DISTCLEAN-distclean}|g
    s|@[A-Z]*_MK@|Makefile|g
    s| *\$||
"
```

If I understand your comment, @Ericson2314's, it looks like optipng's build system is doing precisely what you describe: `    s|@LD@|${LD-\$(CC)}|g`

So when `$LD` isn't set, it defaults to configuring the `Makefile` such that it uses `$CC` as the linker, and I suppose the theory is that they depend on behavior available when invoking `ld` through e.g. `gcc` (though I don't know precisely what that may be).

It would seem reasonable take the `export LD=$CC` (on a per package basis) approach as a solution. I think this can safely be considered a bug in the upstream packages' build systems.

**Comment by Ericson2314** [comment]
> (though I don't know precisely what that may be)

Probably the use of `-Wl,` to pass actual ld flags, e.g. `ld -o foo -lbar -Wl,--gc-section`.

**Comment by the-kenny** [comment]
I implemented the `export LD=$CC` workaround in 9fab349ef5. Feel free to reopen this if you don't think the fix to be adequate.
